[Soup] Replace MIMETypeRegistry::mimeTypeForPath for known resources

spenap opened the pull request:
#### a8634fd4013dd0d7abfe852f3d51b6a0c14f7bc7
<pre>
[Soup] Replace MIMETypeRegistry::mimeTypeForPath for known resources
<a href="https://bugs.webkit.org/show_bug.cgi?id=298096">https://bugs.webkit.org/show_bug.cgi?id=298096</a>

Reviewed by Fujii Hironori and Adrian Perez de Castro.

297585@main started using MIMETypeRegistry::mimeTypeForPath to look up known
resources rather than relying on g_content_type_guess, in order not to
rely on external dependencies (XDG via the GIO API).

However, it failed to notice that mimeTypeForPath uses
mimeTypeForExtension, which in turn uses XDG&apos;s
xdg_mime_get_mime_type_from_file_name.

This change uses a small type mapping before it falls back to
g_content_type_guess.

* Source/WebCore/loader/soup/ResourceLoaderSoup.cpp:
(WebCore::contentTypeLookUpForKnownResource):
(WebCore::ResourceLoader::loadGResource):

Canonical link: <a href="https://commits.webkit.org/299491@main">https://commits.webkit.org/299491@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c6262dead4bd34b57fd1cc69b540a5b3bd3bfc89

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119205 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38887 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29538 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125435 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71268 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121083 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39582 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47468 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/90560 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122158 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31556 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106853 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70970 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30600 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24968 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69081 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101002 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25153 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128446 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46112 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34852 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99116 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46479 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103069 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98896 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25139 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/44376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22365 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45982 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51662 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45449 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48797 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47139 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/loader/soup/ResourceLoaderSoup.cpp:
```diff
@@ -34,14 +34,45 @@
 
+static std::optional<String> contentTypeLookUpForKnownResource(const char* filename)
+{
+    String fileName = String::fromLatin1(filename);
```
I think we don't need to allocate a String here. We can use StringView::fromLatin1 here.

spenap replied:
Good idea, let me check

fujii commented on Source/WebCore/loader/soup/ResourceLoaderSoup.cpp:
```diff
@@ -34,14 +34,45 @@
+    if (dotIndex != notFound && dotIndex < (fileName.length() - 1))
+        extension = StringView(fileName).substring(dotIndex);
+
+    static constexpr std::pair<ComparableLettersLiteral, ASCIILiteral> extensionTypeMapping[] = {
```
I think we should search case insensitive. Use ComparableCaseFoldingASCIILiteral instead of ComparableLettersLiteral?

spenap replied:
We're in control of the files we're checking (they are bundled in), so we can make sure they have the right extension, so I will leave this one as it is.

spenap committed:
[Soup] Replace MIMETypeRegistry::mimeTypeForPath for known resources
https://bugs.webkit.org/show_bug.cgi?id=298096

Reviewed by Fujii Hironori and Adrian Perez de Castro.

297585@main started using MIMETypeRegistry::mimeTypeForPath to look up known
resources rather than relying on g_content_type_guess, in order not to
rely on external dependencies (XDG via the GIO API).

However, it failed to notice that mimeTypeForPath uses
mimeTypeForExtension, which in turn uses XDG's
xdg_mime_get_mime_type_from_file_name.

This change uses a small type mapping before it falls back to
g_content_type_guess.

* Source/WebCore/loader/soup/ResourceLoaderSoup.cpp:
(WebCore::contentTypeLookUpForKnownResource):
(WebCore::ResourceLoader::loadGResource):

Canonical link: https://commits.webkit.org/299491@main
